lib/libc/gen/sysconf.3
.\"	$NetBSD$
.\"
.\" Copyright (c) 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	@(#)sysconf.3	8.3 (Berkeley) 4/19/94
.\"
.Dd February 28, 2016
.Dt SYSCONF 3
.Os
.Sh NAME
.Nm sysconf
.Nd get configurable system variables
.Sh LIBRARY
.Lb libc
.Sh SYNOPSIS
.In unistd.h
.Ft long
.Fn sysconf "int name"
.Sh DESCRIPTION
This interface is defined by
.St -p1003.1-88 .
A far more complete interface is available using
.Xr sysctl 3 .
.Pp
The
.Fn sysconf
function provides a method for applications to determine the current
value of a configurable system limit or option variable.
The
.Fa name
argument specifies the system variable to be queried.
Symbolic constants for each name value are found in the include file
.In unistd.h .
.Pp
The available values are as follows:
.Bl -tag -width "123456"
.It Li _SC_ARG_MAX
The maximum bytes of argument to
.Xr execve 2 .
.It Li _SC_ATEXIT_MAX
The maximum number of functions that may be registered with
.Xr atexit 3 .
.It Li _SC_BARRIERS
The version of
.St -p1003.1
and its
Barriers
option to which the system attempts to conform,
otherwise \-1.
.It Li _SC_CLOCK_SELECTION
Return the
.Tn POSIX
version the implementation of the Clock Selection Option
on this system conforms to,
or \-1 if unavailable.
.It Li _SC_CHILD_MAX
The maximum number of simultaneous processes per user id.
.It Li _SC_CLK_TCK
The number of clock ticks per second.
.It Li _SC_FSYNC
Return 1 if the File Synchronization Option is available on this system,
otherwise \-1.
.It Li _SC_HOST_NAME_MAX
The maximum size of a hostname, including
.Dv NULL .
.It Li _SC_IOV_MAX
The maximum number of
.Va iovec
structures that a process has available for use with
.Xr preadv 2 ,
.Xr pwritev 2 ,
.Xr readv 2 ,
.Xr recvmsg 2 ,
.Xr sendmsg 2
or
.Xr writev 2 .
.It Li _SC_JOB_CONTROL
Return 1 if job control is available on this system, otherwise \-1.
.It Li _SC_LOGIN_NAME_MAX
Returns the size of the storage required for a login name, in bytes,
including the terminating NUL.
.It Li _SC_MAPPED_FILES
Return 1 if the Memory Mapped Files Option is available on this system,
otherwise \-1.
.It Li _SC_MEMLOCK
Return 1 if the Process Memory Locking Option is available on this system,
otherwise \-1.
.It Li _SC_MEMLOCK_RANGE
Return 1 if the Range Memory Locking Option is available on this system,
otherwise \-1.
.It Li _SC_MEMORY_PROTECTION
Return 1 if the Memory Protection Option is available on this system,
otherwise \-1.
.It Li _SC_MONOTONIC_CLOCK
Return the
.Tn POSIX
version the implementation of the Monotonic Clock Option
on this system conforms to,
or \-1 if unavailable.
.It Li _SC_NGROUPS_MAX
The maximum number of supplemental groups.
.It Li _SC_OPEN_MAX
The maximum number of open files per process.
.It Li _SC_PAGESIZE
The size of a system page in bytes.
.It Li _SC_PASS_MAX
The maximum length of the password, not counting
.Dv NULL .
.It Li _SC_READER_WRITER_LOCKS
The version of
.St -p1003.1
and its
Read-Write Locks
option to which the system attempts to conform,
otherwise \-1.
.It Li _SC_REGEXP
Return 1 if
.Tn POSIX
regular expressions are available on this system, otherwise \-1.
.It Li _SC_SEMAPHORES
The version of
.St -p1003.1
and its
Semaphores
option to which the system attempts to conform,
otherwise \-1.
.It Li _SC_SHELL
Return 1 if
.Tn POSIX
shell is available on this system, otherwise \-1.
.It Li _SC_SPIN_LOCKS
The version of
.St -p1003.1
and its
Spin Locks
option to which the system attempts to conform,
otherwise \-1.
.It Li _SC_STREAM_MAX
The minimum maximum number of streams that a process may have open
at any one time.
.It Li _SC_SYMLOOP_MAX
The maximum number of symbolic links that may be expanded in a path name.
.It Li _SC_SYNCHRONIZED_IO
Return 1 if the Synchronized I/O Option is available on this system,
otherwise \-1.
.It Li _SC_THREADS
The version of
.St -p1003.1
and its
Threads
option to which the system attempts to conform,
otherwise \-1.
.It Li _SC_TIMERS
The version of
.St -p1003.1
and its
Timers
option to which the system attempts to conform,
otherwise \-1.
.It Li _SC_TZNAME_MAX
The minimum maximum number of types supported for the name of a
timezone.
.It Li _SC_SAVED_IDS
Returns 1 if saved set-group and saved set-user ID is available,
otherwise \-1.
.It Li _SC_VERSION
The version of ISO/IEC 9945 (POSIX 1003.1) with which the system
attempts to comply.
.It Li _SC_XOPEN_SHM
Return 1 if the
.St -xpg4.2
Shared Memory
option is available on this system,
otherwise \-1.
.Pp
Availability of the
Shared Memory
option depends on the
.Li SYSVSHM
kernel option.
.It Li _SC_BC_BASE_MAX
The maximum ibase/obase values in the
.Xr bc 1
utility.
.It Li _SC_BC_DIM_MAX
The maximum array size in the
.Xr bc 1
utility.
.It Li _SC_BC_SCALE_MAX
The maximum scale value in the
.Xr bc 1
utility.
.It Li _SC_BC_STRING_MAX
The maximum string length in the
.Xr bc 1
utility.
.It Li _SC_COLL_WEIGHTS_MAX
The maximum number of weights that can be assigned to any entry of
the LC_COLLATE order keyword in the locale definition file.
.It Li _SC_EXPR_NEST_MAX
The maximum number of expressions that can be nested within
parenthesis by the
.Xr expr 1
utility.
.It Li _SC_LINE_MAX
The maximum length in bytes of a text-processing utility's input
line.
.It Li _SC_RE_DUP_MAX
The maximum number of repeated occurrences of a regular expression
permitted when using interval notation.
.It Li _SC_2_VERSION
The version of POSIX 1003.2 with which the system attempts to comply.
.It Li _SC_2_C_BIND
Return 1 if the system's C-language development facilities support the
C-Language Bindings Option, otherwise \-1.
.It Li _SC_2_C_DEV
Return 1 if the system supports the C-Language Development Utilities Option,
otherwise \-1.
.It Li _SC_2_CHAR_TERM
Return 1 if the system supports at least one terminal type capable of
all operations described in POSIX 1003.2, otherwise \-1.
.It Li _SC_2_FORT_DEV
Return 1 if the system supports the FORTRAN Development Utilities Option,
otherwise \-1.
.It Li _SC_2_FORT_RUN
Return 1 if the system supports the FORTRAN Runtime Utilities Option,
otherwise \-1.
.It Li _SC_2_LOCALEDEF
Return 1 if the system supports the creation of locales, otherwise \-1.
.It Li _SC_2_SW_DEV
Return 1 if the system supports the Software Development Utilities Option,
otherwise \-1.
.It Li _SC_2_UPE
Return 1 if the system supports the User Portability Utilities Option,
otherwise \-1.
.It Li _SC_GETGR_R_SIZE_MAX
The minimum size of the
.Fa buffer
passed to
.Xr getgrgid_r 3
and
.Xr getgrnam_r 3 .
.It Li _SC_GETPW_R_SIZE_MAX
The minimum size of the
.Fa buffer
passed to
.Xr getpwnam_r 3
and
.Xr getpwuid_r 3 .
.It Li _SC_NPROCESSORS_CONF
The number of processors configured.
.It Li _SC_NPROCESSORS_ONLN
The number of processors online (capable of running processes).
.It Li _SC_PHYS_PAGES
The amount of physical memory on the system in
.Li _SC_PAGESIZE
bytes.
.Li _SC_TIMER_MAX
The number of timers available for
.Xr timer_create 2 .
This is also known as
.Dv _POSIX_TIMER_MAX .
.El
.Sh RETURN VALUES
If the call to
.Nm sysconf
is not successful, \-1 is returned and
.Va errno
is set appropriately.
Otherwise, if the variable is associated with functionality that is not
supported, \-1 is returned and
.Va errno
is not modified.
Otherwise, the current variable value is returned.
.Sh ERRORS
The
.Fn sysconf
function may fail and set
.Va errno
for any of the errors specified for the library functions
.Xr sysctl 3 .
In addition, the following error may be reported:
.Bl -tag -width Er
.It Bq Er EINVAL
The value of the
.Fa name
argument is invalid.
.El
.Sh SEE ALSO
.Xr getconf 1 ,
.Xr limits 3 ,
.Xr sysctl 3
.Sh STANDARDS
The
.Fn sysconf
function conforms to
.St -p1003.1-90 .
The constants
.Li _SC_NPROCESSORS_CONF
and
.Li _SC_NPROCESSORS_ONLN
are not part of the standard, but are provided by many systems.
.Sh HISTORY
The
.Nm sysconf
function first appeared in
.Bx 4.4 .
.Sh BUGS
The value for _SC_STREAM_MAX is a minimum maximum, and required to be
the same as ANSI C's FOPEN_MAX, so the returned value is a ridiculously
small and misleading number.


doc/CHANGES
# LIST OF CHANGES FROM LAST RELEASE:			<$Revision$>
#
#
# [Note: This file does not mention every change made to the NetBSD source tree.
# For an exhaustive list, please subscribe to the `source-changes' mailing list,
# or see the mailing list archives.  For more information, send a message
# containing just the word `help' to `majordomo@NetBSD.org'.]
#
# ----------------------------------------------------------------------------
# PLEASE USE THE FOLLOWING STYLE WHEN ADDING ITEMS:
#
#	theme: Content ...
#		more content ...
#		more content. [developer DATE-FORMAT]
#
# That's it, one tab to start an item, and two tabs to indent the content.
# Also please start the content with a capital and end it with a dot.
#
# This is very important for the www scripts, otherwise it won't be able to
# parse the file and the HTML output will be mangled.
#
# See htutils/changes/changes2html script for more details.
#
# ----------------------------------------------------------------------------


Changes from NetBSD 7.0 to NetBSD 8.0:
	libc: Update to tzcode2014f. [christos 20140815]
	gmake:	Update to gmake-3.81 in external/gpl2/gmake and remove 
		gnu/dist/gmake (3.80) [christos 20140818]
	module(7): Create MODULAR_DEFAULT_AUTOLOAD config flag to control
		whether module autoloading is on by default.
		[jnemeth 20140824]
	kernel: Option SYMTAB_SPACE has been replaced with COPY_SYMTAB.
		The related changes to src/sys/conf require a rebuild of
		config(1). [joerg 20140824]
	zoneinfo: Import tzdata2014g. [apb 20140831]
	wm(4): Add internal SERDES support for 82575 and newer controllers.
		[msaitoh 20140903]
	or1k: Add new port.  [matt 20140903]
	awindma(4): Add support for Allwinner A10/A20 DMA controller.
		[jmcneill 20140906]
	awinac(4): Add support for Allwinner A10/A20 audio codec.
		[jmcneill 20140906]
	awinrtc(4): Add support for Allwinner A10/A20 real-time clock.
		[jmcneill 20140907]
	awinmmc(4): Use DMA for MMC transfers. [jmcneill 20140908]
	axp20x(4): Add AXP20x PMU driver. [jmcneill 20140909]
	rtsol(8): Removed in favour of dhcpcd. [roy 20140911]
	rtsold(8): Removed in favour of dhcpcd. [roy 20140911]
	dhcpcd(8): Import dhcpcd-6.4.5. [roy 20140918]
	arm: Add support for i.MX6 SoC. [ryo 20140925]
	gpt(8): Completed overhaul, including adding follwing subcommands:
		resize, set, unset, backup, restore, and resizedisk.
		[jnemeth 20140926]
	dhcpcd(8): Import dhcpcd-6.4.7. [roy 20140927]
	mpt(4): added bio(4) support. [jmcneill 20140927]
	gpt(8): added type subcommand to modify partition type.
		[jnemeth 20140927]
	zoneinfo: Import tzdata2014h. [apb 20141004]
	dhcpcd(8): Import dhcpcd-6.5.0. [roy 20141006]
	libc: Update to tzcode2014h. [christos 20141007]
	file(1): Upgraded to 5.20. [christos 20141010]
	arm: Add support for Allwinner A31 SoC. [jmcneill 20141010]
	awiniic(4): Add support for Allwinner A31 SoC. [jmcneill 20141012]
	awinp2wi(4): Add support for Allwinner A31 Push-Pull Two Wire
		Interface. [jmcneill 20141012]
	axp22x(4): Add AXP22x PMU driver. [jmcneill 20141012]
	awinrtc(4): Add support for Allwinner A31 real-time clock.
		[jmcneill 20141012]
	awindma(4): Add support for Allwinner A31 DMA controller.
		[jmcneill 20141013]
	awinac(4): Add support for Allwinner A31 audio codec.
		[jmcneill 20141013]
	motg(4): Add support for Allwinner A10/A20 SoC. [jmcneill 20141015]
	wpa: Import wpa_supplicant and hostapd 2.3. [christos 20141016]
	dhcpcd(8): Import dhcpcd-6.5.1. [roy 20141018]
	OpenSSH: Imported 6.7. [christos 20141018]
	resolvconf(8): Import openresolv-3.6.0 [roy 20141020]
	gcc: Support -fsanitize=address [christos 20141022]
	zoneinfo: Import tzdata2014i. [apb 20141023]
	libc: Update to tzcode2014i. [christos 20141023]
	openpam(3): update to 20140912 (ourouparia) [christos 20141024]
	pppd(8): updated to version 2.4.7. [christos 20141025]
	acpi(4): Updated ACPICA to 20140926. [christos 20141025]
	dhcpcd(8): Import dhcpcd-6.6.0 [roy 20141029]
	resolvconf(8): Import openresolv-3.6.1 [roy 20141029]
	rtsx(4): Add support for Realtek RTS5227/RTL8402/RTL8411/RTL8411B.
		[nonaka 20141029]
	flex(1): Import flex-2.5.37 [christos 20141029]
	iwn(4): Add support for Intel Centrino Wireless-N
		100/105/130/135/2200/2230. From OpenBSD. [nonaka 20141030]
	awge(4): Add support for gigabit ethernet found on various
		Allwinner SOCs. [martin 20141101]
	dhcpcd(8): Import dhcpcd-6.6.1. [roy 20141107]
	awinhdmi(4): Add support for Allwinner A20/A31 HDMI controller.
		[jmcneill 20141110]
	awintcon(4): Add support for Allwinner A20/A31 TV/LCD controller.
		[jmcneill 20141110]
	awindebe(4): Add support for Allwinner A20/A31 Display engine
		backend (DE-BE). [jmcneill 20141110]
	awinhdmiaudio(4): Add support for Allwinner A20/A31 HDMI audio
		controller. [jmcneill 20141111]
	dhcpcd(8): Import dhcpcd-6.6.2. [roy 20141114]
	awinir(4): Add support for Allwinner A20/A31 IR receiver.
		[jmcneill 20141115]
	zoneinfo: Import tzdata2014j. [apb 20141116]
	libc: Change arc4random(3) to use ChaCha20 and per-thread state.
		[riastradh 20141116]
	libc: Update to tzcode2014j. [christos 20141117]
	libpcap: Import 1.6.2. [christos 20141119]
	tcpdump(8): Import 4.6.2. [christos 20141119]
	dhcpcd(8): Import dhcpcd-6.6.4. [roy 20141126]
	dhcpcd(8): Import dhcpcd-6.6.5. [roy 20141209]
	bind: Import version 9.10.1-P1. [christos 20141209]
	dhcpcd(8): Import dhcpcd-6.6.6. [roy 20141217]
	powerpc: Switch to GCC 4.8. [matt 20141218]
	dhcpcd(8): Import dhcpcd-6.6.7. [roy 20141219]
	ntp: Import ntp 4.2.8. [christos 20141219]
	midirecord(1): Add program to record MIDI files.  [mrg 20141229]
	file: update to 5.22 security limits [christos 20150102]
	byacc: update to 20141128 for more bison compat [christos 20150103]
	mgx: driver for Southland Media Systems (now Quantum 3D) MGX SBus
		graphics devices [macallan 20150106]
	gcc: Import GCC 4.8.4.  [mrg 20150109]
	openssl: Import openssl 1.0.1k [spz 20150113]
	am-utils: Upgrade to 6.2; adds nfsv4, lustre (which we can't use)
		nfs3 rpc support for amfs [christos 20150117]

lib/libc/gen/sysconf.c
/*	$NetBSD$	*/

/*-
 * Copyright (c) 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * This code is derived from software contributed to Berkeley by
 * Sean Eric Fagan of Cygnus Support.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/cdefs.h>
#if defined(LIBC_SCCS) && !defined(lint)
#if 0
static char sccsid[] = "@(#)sysconf.c	8.2 (Berkeley) 3/20/94";
#else
__RCSID("$NetBSD$");
#endif
#endif /* LIBC_SCCS and not lint */

#include "namespace.h"
#include <sys/param.h>
#include <sys/sysctl.h>
#include <sys/time.h>
#include <sys/resource.h>

#include <errno.h>
#include <limits.h>
#include <time.h>
#include <unistd.h>
#include <paths.h>
#include <pwd.h>

#ifdef __weak_alias
__weak_alias(sysconf,__sysconf)
#endif

/*
 * sysconf --
 *	get configurable system variables.
 *
 * XXX
 * POSIX 1003.1 (ISO/IEC 9945-1, 4.8.1.3) states that the variable values
 * not change during the lifetime of the calling process.  This would seem
 * to require that any change to system limits kill all running processes.
 * A workaround might be to cache the values when they are first retrieved
 * and then simply return the cached value on subsequent calls.  This is
 * less useful than returning up-to-date values, however.
 */
long
sysconf(int name)
{
	struct rlimit rl;
	size_t len;
	uint64_t mem;
	int mib[CTL_MAXNAME], value;
	unsigned int mib_len;
	struct clockinfo tmpclock;
	static int clk_tck;

	len = sizeof(value);

	/* Default length of the MIB */
	mib_len = 2;

	switch (name) {

/* 1003.1 */
	case _SC_ARG_MAX:
		mib[0] = CTL_KERN;
		mib[1] = KERN_ARGMAX;
		break;
	case _SC_CHILD_MAX:
		return (getrlimit(RLIMIT_NPROC, &rl) ? -1 : (long)rl.rlim_cur);
	case _O_SC_CLK_TCK:
		/*
		 * For applications compiled when CLK_TCK was a compile-time
		 * constant.
		 */
		return 100;
	case _SC_CLK_TCK:
		/*
		 * Has to be handled specially because it returns a
		 * struct clockinfo instead of an integer. Also, since
		 * this might be called often by some things that
		 * don't grok CLK_TCK can be a macro expanding to a
		 * function, cache the value.
		 */
		if (clk_tck == 0) {
			mib[0] = CTL_KERN;
			mib[1] = KERN_CLOCKRATE;
			len = sizeof(struct clockinfo);
			clk_tck = sysctl(mib, 2, &tmpclock, &len, NULL, 0)
			    == -1 ? -1 : tmpclock.hz;
		}
		return(clk_tck);
	case _SC_JOB_CONTROL:
		mib[0] = CTL_KERN;
		mib[1] = KERN_JOB_CONTROL;
		goto yesno;
	case _SC_NGROUPS_MAX:
		mib[0] = CTL_KERN;
		mib[1] = KERN_NGROUPS;
		break;
	case _SC_OPEN_MAX:
		return (getrlimit(RLIMIT_NOFILE, &rl) ? -1 : (long)rl.rlim_cur);
	case _SC_STREAM_MAX:
		mib[0] = CTL_USER;
		mib[1] = USER_STREAM_MAX;
		break;
	case _SC_TZNAME_MAX:
		mib[0] = CTL_USER;
		mib[1] = USER_TZNAME_MAX;
		break;
	case _SC_SAVED_IDS:
		mib[0] = CTL_KERN;
		mib[1] = KERN_SAVED_IDS;
		goto yesno;
	case _SC_VERSION:
		mib[0] = CTL_KERN;
		mib[1] = KERN_POSIX1;
		break;

/* 1003.1b */
	case _SC_PAGESIZE:
		return _getpagesize();
	case _SC_FSYNC:
		mib[0] = CTL_KERN;
		mib[1] = KERN_FSYNC;
		goto yesno;
	case _SC_SYNCHRONIZED_IO:
		mib[0] = CTL_KERN;
		mib[1] = KERN_SYNCHRONIZED_IO;
		goto yesno;
	case _SC_MAPPED_FILES:
		mib[0] = CTL_KERN;
		mib[1] = KERN_MAPPED_FILES;
		goto yesno;
	case _SC_MEMLOCK:
		mib[0] = CTL_KERN;
		mib[1] = KERN_MEMLOCK;
		goto yesno;
	case _SC_MEMLOCK_RANGE:
		mib[0] = CTL_KERN;
		mib[1] = KERN_MEMLOCK_RANGE;
		goto yesno;
	case _SC_MEMORY_PROTECTION:
		mib[0] = CTL_KERN;
		mib[1] = KERN_MEMORY_PROTECTION;
		goto yesno;
	case _SC_MONOTONIC_CLOCK:
		mib[0] = CTL_KERN;
		mib[1] = KERN_MONOTONIC_CLOCK;
		goto yesno;
	case _SC_SEMAPHORES:
		mib[0] = CTL_KERN;
		mib[1] = KERN_POSIX_SEMAPHORES;
		goto yesno;
	case _SC_TIMERS:
		mib[0] = CTL_KERN;
		mib[1] = KERN_POSIX_TIMERS;
		goto yesno;

/* 1003.1c */
	case _SC_LOGIN_NAME_MAX:
		mib[0] = CTL_KERN;
		mib[1] = KERN_LOGIN_NAME_MAX;
		break;
	case _SC_THREADS:
		mib[0] = CTL_KERN;
		mib[1] = KERN_POSIX_THREADS;
		goto yesno;

/* 1003.1j */
	case _SC_BARRIERS:
		mib[0] = CTL_KERN;
		mib[1] = KERN_POSIX_BARRIERS;
		goto yesno;
	case _SC_SPIN_LOCKS:
		mib[0] = CTL_KERN;
		mib[1] = KERN_POSIX_SPIN_LOCKS;
		goto yesno;
	/* Historical; Threads option in 1003.1-2001 */
	case _SC_READER_WRITER_LOCKS:
		mib[0] = CTL_KERN;
		mib[1] = KERN_POSIX_READER_WRITER_LOCKS;
		goto yesno;

/* 1003.2 */
	case _SC_BC_BASE_MAX:
		mib[0] = CTL_USER;
		mib[1] = USER_BC_BASE_MAX;
		break;
	case _SC_BC_DIM_MAX:
		mib[0] = CTL_USER;
		mib[1] = USER_BC_DIM_MAX;
		break;
	case _SC_BC_SCALE_MAX:
		mib[0] = CTL_USER;
		mib[1] = USER_BC_SCALE_MAX;
		break;
	case _SC_BC_STRING_MAX:
		mib[0] = CTL_USER;
		mib[1] = USER_BC_STRING_MAX;
		break;
	case _SC_COLL_WEIGHTS_MAX:
		mib[0] = CTL_USER;
		mib[1] = USER_COLL_WEIGHTS_MAX;
		break;
	case _SC_EXPR_NEST_MAX:
		mib[0] = CTL_USER;
		mib[1] = USER_EXPR_NEST_MAX;
		break;
	case _SC_LINE_MAX:
		mib[0] = CTL_USER;
		mib[1] = USER_LINE_MAX;
		break;
	case _SC_RE_DUP_MAX:
		mib[0] = CTL_USER;
		mib[1] = USER_RE_DUP_MAX;
		break;
	case _SC_2_VERSION:
		mib[0] = CTL_USER;
		mib[1] = USER_POSIX2_VERSION;
		break;
	case _SC_2_C_BIND:
		mib[0] = CTL_USER;
		mib[1] = USER_POSIX2_C_BIND;
		goto yesno;
	case _SC_2_C_DEV:
		mib[0] = CTL_USER;
		mib[1] = USER_POSIX2_C_DEV;
		goto yesno;
	case _SC_2_CHAR_TERM:
		mib[0] = CTL_USER;
		mib[1] = USER_POSIX2_CHAR_TERM;
		goto yesno;
	case _SC_2_FORT_DEV:
		mib[0] = CTL_USER;
		mib[1] = USER_POSIX2_FORT_DEV;
		goto yesno;
	case _SC_2_FORT_RUN:
		mib[0] = CTL_USER;
		mib[1] = USER_POSIX2_FORT_RUN;
		goto yesno;
	case _SC_2_LOCALEDEF:
		mib[0] = CTL_USER;
		mib[1] = USER_POSIX2_LOCALEDEF;
		goto yesno;
	case _SC_2_SW_DEV:
		mib[0] = CTL_USER;
		mib[1] = USER_POSIX2_SW_DEV;
		goto yesno;
	case _SC_2_UPE:
		mib[0] = CTL_USER;
		mib[1] = USER_POSIX2_UPE;
		goto yesno;

/* XPG 4.2 */
	case _SC_IOV_MAX:
		mib[0] = CTL_KERN;
		mib[1] = KERN_IOV_MAX;
		break;
	case _SC_XOPEN_SHM:
		mib[0] = CTL_KERN;
		mib[1] = KERN_SYSVIPC;
		mib[2] = KERN_SYSVIPC_SHM;
		mib_len = 3;
		goto yesno;

/* 1003.1-2001, XSI Option Group */
	case _SC_AIO_LISTIO_MAX:
		if (sysctlgetmibinfo("kern.aio_listio_max", &mib[0], &mib_len,
		    NULL, NULL, NULL, SYSCTL_VERSION))
			return -1;
		break; 
	case _SC_AIO_MAX:
		if (sysctlgetmibinfo("kern.aio_max", &mib[0], &mib_len,
		    NULL, NULL, NULL, SYSCTL_VERSION))
			return -1;
		break; 
	case _SC_ASYNCHRONOUS_IO:
		if (sysctlgetmibinfo("kern.posix_aio", &mib[0], &mib_len,
		    NULL, NULL, NULL, SYSCTL_VERSION))
			return -1;
		goto yesno;
	case _SC_MESSAGE_PASSING:
		if (sysctlgetmibinfo("kern.posix_msg", &mib[0], &mib_len,
		    NULL, NULL, NULL, SYSCTL_VERSION))
			return -1;
		goto yesno;
	case _SC_MQ_OPEN_MAX:
		if (sysctlgetmibinfo("kern.mqueue.mq_open_max", &mib[0],
		    &mib_len, NULL, NULL, NULL, SYSCTL_VERSION))
			return -1;
		break; 
	case _SC_MQ_PRIO_MAX:
		if (sysctlgetmibinfo("kern.mqueue.mq_prio_max", &mib[0],
		    &mib_len, NULL, NULL, NULL, SYSCTL_VERSION))
			return -1;
		break; 
	case _SC_PRIORITY_SCHEDULING:
		if (sysctlgetmibinfo("kern.posix_sched", &mib[0], &mib_len,
		    NULL, NULL, NULL, SYSCTL_VERSION))
			return -1;
		goto yesno;
	case _SC_ATEXIT_MAX:
		mib[0] = CTL_USER;
		mib[1] = USER_ATEXIT_MAX;
		break;

/* 1003.1-2001, TSF */
	case _SC_GETGR_R_SIZE_MAX:
		return _GETGR_R_SIZE_MAX;
	case _SC_GETPW_R_SIZE_MAX:
		return _GETPW_R_SIZE_MAX;

/* Unsorted */
	case _SC_HOST_NAME_MAX:
		return MAXHOSTNAMELEN;
	case _SC_PASS_MAX:
		return _PASSWORD_LEN;
	case _SC_REGEXP:
		return _POSIX_REGEXP;
	case _SC_SHARED_MEMORY_OBJECTS:
		return _POSIX_SHARED_MEMORY_OBJECTS;
	case _SC_SHELL:
		return _POSIX_SHELL;
	case _SC_SPAWN:
		return _POSIX_SPAWN;
	case _SC_SYMLOOP_MAX:
		return MAXSYMLINKS;

yesno:		if (sysctl(mib, mib_len, &value, &len, NULL, 0) == -1)
			return (-1);
		if (value == 0)
			return (-1);
		return (value);

/* Extensions */
	case _SC_NPROCESSORS_CONF:
		mib[0] = CTL_HW;
		mib[1] = HW_NCPU;
		break;
	case _SC_NPROCESSORS_ONLN:
		mib[0] = CTL_HW;
		mib[1] = HW_NCPUONLINE;
		break;

/* Linux/Solaris */
	case _SC_PHYS_PAGES:
		len = sizeof(mem);
		mib[0] = CTL_HW;
		mib[1] = HW_PHYSMEM64;
		return sysctl(mib, 2, &mem, &len, NULL, 0) == -1 ? -1 : 
		    (long)(mem / _getpagesize()); 

/* Native */
	case _SC_SCHED_RT_TS:
		if (sysctlgetmibinfo("kern.sched.rtts", &mib[0], &mib_len,
		    NULL, NULL, NULL, SYSCTL_VERSION))      
			return -1;              
		break;
	case _SC_SCHED_PRI_MIN:
		if (sysctlgetmibinfo("kern.sched.pri_min", &mib[0], &mib_len,
		    NULL, NULL, NULL, SYSCTL_VERSION))
			return -1;
		break;
	case _SC_SCHED_PRI_MAX:
		if (sysctlgetmibinfo("kern.sched.pri_max", &mib[0], &mib_len,
		    NULL, NULL, NULL, SYSCTL_VERSION))
			return -1;
		break;
	case _SC_THREAD_DESTRUCTOR_ITERATIONS:
		return _POSIX_THREAD_DESTRUCTOR_ITERATIONS;
	case _SC_THREAD_KEYS_MAX:
		return _POSIX_THREAD_KEYS_MAX;
	case _SC_THREAD_STACK_MIN:
		return _getpagesize();
	case _SC_THREAD_THREADS_MAX:
		if (sysctlgetmibinfo("kern.maxproc", &mib[0], &mib_len,
		    NULL, NULL, NULL, SYSCTL_VERSION))	/* XXX */
			return -1;
		goto yesno;
	case _SC_THREAD_ATTR_STACKADDR:
		return _POSIX_THREAD_ATTR_STACKADDR;
	case _SC_THREAD_ATTR_STACKSIZE:
		return _POSIX_THREAD_ATTR_STACKSIZE;
	case _SC_THREAD_SAFE_FUNCTIONS:
		return _POSIX_THREAD_SAFE_FUNCTIONS;
	case _SC_THREAD_PRIORITY_SCHEDULING:
	case _SC_THREAD_PRIO_INHERIT:
	case _SC_THREAD_PRIO_PROTECT:
	case _SC_THREAD_PROCESS_SHARED:
		return -1;
	case _SC_TTY_NAME_MAX:
		return pathconf(_PATH_DEV, _PC_NAME_MAX);
	case _SC_TIMER_MAX:
		return _POSIX_TIMER_MAX;
	default:
		errno = EINVAL;
		return (-1);
	}
	return (sysctl(mib, mib_len, &value, &len, NULL, 0) == -1 ? -1 : value); 
}
